File: gitdot-web/packages/gitdot-api/src/endpoint/user/list-user-repositories.ts
import { z } from "zod";
import { page, RepositoryResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const ListUserRepositoriesRequest = z.object({
  cursor: z.string().optional(),
  limit: z.number().int().positive().optional(),
});
export type ListUserRepositoriesRequest = z.infer<
  typeof ListUserRepositoriesRequest
>;

export const ListUserRepositoriesResponse = page(RepositoryResource);
export type ListUserRepositoriesResponse = z.infer<
  typeof ListUserRepositoriesResponse
>;

export const ListUserRepositories = {
  path: "/user/{user_name}/repositories",
  method: "GET",
  request: ListUserRepositoriesRequest,
  response: ListUserRepositoriesResponse,
} as const satisfies Endpoint;
export type ListUserRepositories = typeof ListUserRepositories;

File: gitdot-api/src/endpoint/user/list_user_repositories.rs
use serde::{Deserialize, Serialize};

use crate::{
    endpoint::Endpoint,
    resource::{common::Page, user::UserRepositoryResource},
};

pub struct ListUserRepositories;

impl Endpoint for ListUserRepositories {
    const PATH: &'static str = "/user/{user_name}/repositories";
    const METHOD: http::Method = http::Method::GET;

    type Request = ListUserRepositoriesRequest;
    type Response = ListUserRepositoriesResponse;
}

#[derive(ApiRequest, Debug, Default, Serialize, Deserialize)]
pub struct ListUserRepositoriesRequest {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cursor: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub limit: Option<u32>,
}

pub type ListUserRepositoriesResponse = Page<UserRepositoryResource>;
